pandas/_libs/custom_window.py

```diff
@@ -60,3 +56,295 @@ def get_window_bounds(
+    right_closed: bint
+        right endpoint closedness
+        True if the right endpoint is closed, False if open
+    floor: optional
+        unit for flooring the unit
+    """
+    def __init__(self,
+                 values: np.ndarray,
+                 win: int,
+                 minp: int,
+                 left_closed: bool,
+                 right_closed: bool,
+                 index: np.ndarray,
+                 floor=None):
+
+        self.is_variable = 1
+        self.N = len(index)
+
+        self.start = np.empty(self.N, dtype='int64')
+        self.start.fill(-1)
+
+        self.end = np.empty(self.N, dtype='int64')
+        self.end.fill(-1)
+
+        self.build(index, win, left_closed, right_closed)
+
+        # max window size
+        self.win = (self.end - self.start).max()
+
+    def build(self,
+              index: np.ndarray,
+              win: int,
+              left_closed: bool,
+              right_closed: bool):
+
+        start = self.start
+        end = self.end
+        N = self.N
+
+        start[0] = 0
+
+        # right endpoint is closed
+        if right_closed:
+            end[0] = 1
+        # right endpoint is open
+        else:
+            end[0] = 0
+
+        # start is start of slice interval (including)
+        # end is end of slice interval (not including)
+        for i in range(1, N):
+            end_bound = index[i]
+            start_bound = index[i] - win
+
+            # left endpoint is closed
+            if left_closed:
+                start_bound -= 1
+
+            # advance the start bound until we are
+            # within the constraint
+            start[i] = i
+            for j in range(start[i - 1], i):
+                if index[j] > start_bound:
+                    start[i] = j
+                    break
+
+            # end bound is previous end
+            # or current index
+            if index[end[i - 1]] <= end_bound:
+                end[i] = i + 1
+            else:
+                end[i] = end[i - 1]
+
+            # right endpoint is open
+            if not right_closed:
+                end[i] -= 1
+
+
+def get_window_indexer(values: np.ndarray, win: int, index: Optional[np.ndarray],
+                       closed: Optional[str], floor=None, use_mock: bool = True,
+                       custom_indexer=None) -> BeginEnd:
+    """
+    Return a tuple of 2, 1D arrays. The first array contains the starting window indices and
+    and the second array contains the ending window indices.
+
+    Parameters
+    ----------
+    values: 1d ndarray
+    win: integer, window size
+    index: 1d ndarray, optional
+        index to the values array
+    closed: string, default None
+        {'right', 'left', 'both', 'neither'}
+        window endpoint closedness. Defaults to 'right' in
+        VariableWindowIndexer and to 'both' in FixedWindowIndexer
+    floor: optional
+        unit for flooring the unit
+    use_mock: boolean, default True
```

Comment:
do we need to expose the mock? what is this for?

Reply:
It looks to be so that a different window bound calculation happens for median and quantile routines. I imagine it _may_ not be entirely necessary anymore.